File: Design Patterns/Structural/Facade.ts
class RoomService {
  checkAvailability(roomType: string) {
    console.log(`Checking availability for ${roomType}`)
    return true
  }
}

class PaymentService {
  processPayment(amount: number) {
    console.log(`Processing payment of $${amount}`)
  }
}

class NotificationService {
  sendConfirmation() {
    console.log("Booking confirmation sent")
  }
}

class HotelBookingFacade {
  private roomService = new RoomService()
  private paymentService = new PaymentService()
  private notificationService = new NotificationService()

  bookRoom(roomType: string, amount: number) {
    if (this.roomService.checkAvailability(roomType)) {
      this.paymentService.processPayment(amount)
      this.notificationService.sendConfirmation()
      console.log("Room booked successfully")
    }
  }
}

const booking = new HotelBookingFacade()
booking.bookRoom("Deluxe", 200)

File: C#/Design Patterns/Structural/Facade.cs
using System;

class RoomService
{
    public bool CheckAvailability(string roomType)
    {
        Console.WriteLine($"Checking availability for {roomType}");
        return true;
    }
}

class PaymentService
{
    public void ProcessPayment(int amount)
    {
        Console.WriteLine($"Processing payment of {amount}");
    }
}

class NotificationService
{
    public void SendConfirmation()
    {
        Console.WriteLine("Booking confirmation sent");
    }
}

class HotelBookingFacade
{
    private readonly RoomService _roomService = new RoomService();
    private readonly PaymentService _paymentService = new PaymentService();
    private readonly NotificationService _notificationService = new NotificationService();

    public void BookRoom(string roomType, int amount)
    {
        if (_roomService.CheckAvailability(roomType))
        {
            _paymentService.ProcessPayment(amount);
            _notificationService.SendConfirmation();
            Console.WriteLine("Room booked successfully");
        }
    }
}

class Program
{
    public static void Main(string[] args)
    {
        HotelBookingFacade booking = new HotelBookingFacade();
        booking.BookRoom("Deluxe", 200);
    }
}